pallets/circuit/src/lib.rs

```diff
@@ -648,62 +563,55 @@ pub mod pallet {
-
-            if requester != local_ctx.xtx.requester
-                || local_ctx.xtx.status > CircuitStatus::PendingBidding
-            {
-                return Err(Error::<T>::UnauthorizedCancellation.into())
-            }
+            let attempting_requester = Self::authorize(origin, CircuitRole::Requester)?;
 
-            // Drop cancellation in case some bids have already been posted
-            if Self::get_current_step_fsx(&local_ctx)
-                .iter()
-                .any(|fsx| fsx.best_bid.is_some())
-            {
-                return Err(Error::<T>::UnauthorizedCancellation.into())
-            }
+            Machine::<T>::compile(
+                &mut Machine::<T>::load_xtx(xtx_id)?,
+                |current_fsx, _local_state, _steps_cnt, status, requester| {
+                    if attempting_requester != requester || status > CircuitStatus::PendingBidding {
+                        return Err(Error::<T>::UnauthorizedCancellation)
+                    }
+                    // Drop cancellation in case some bids have already been posted
+                    if current_fsx.iter().any(|fsx| fsx.best_bid.is_some()) {
+                        return Err(Error::<T>::UnauthorizedCancellation)
+                    }
+                    Ok(PrecompileResult::TryKill(Cause::IntentionalKill))
+                },
+                no_post_updates,
+            )?;
 
-            Self::kill(&mut local_ctx, CircuitStatus::DroppedAtBidding);
+            Ok(().into())
+        }
 
+        #[pallet::weight(<T as pallet::Config>::WeightInfo::cancel_xtx())]
+        pub fn revert(origin: OriginFor<T>, xtx_id: T::Hash) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+            let _success =
+                Machine::<T>::revert(xtx_id, Cause::IntentionalKill, infallible_no_post_updates);
             Ok(().into())
         }
 
         #[pallet::weight(<T as pallet::Config>::WeightInfo::on_extrinsic_trigger())]
         pub fn on_extrinsic_trigger(
             origin: OriginFor<T>,
             side_effects: Vec<SideEffect<T::AccountId, BalanceOf<T>>>,
-            sequential: bool,
+            _sequential: bool,
         ) -> DispatchResultWithPostInfo {
             // Authorize: Retrieve sender of the transaction.
             let requester = Self::authorize(origin, CircuitRole::Requester)?;
-            // Setup: new xtx context
-            let mut local_xtx_ctx: LocalXtxCtx<T> =
-                Self::setup(CircuitStatus::Requested, &requester, None)?;
-
-            // Validate: Side Effects
-            Self::validate(&side_effects, &mut local_xtx_ctx, &requester, sequential).map_err(
-                |e| {
-                    log::error!("Self::validate hit an error -- {:?}", e);
-                    Error::<T>::SideEffectsValidationFailed
+            // Setup: new xtx context with SFX validation
+            let mut fresh_xtx = Machine::<T>::setup(&side_effects, &requester)?;
```

Comment:
I think we're missing a critical validation here. We need to ensure that for each SFX the same asset_id is used for max_reward & insurance deposit. Otherwise, we run into issues with slashing. Couldn't find anything in ´fn validate´ and protocol

Reply:
there's a single `reward_asset_id` always for both insurance and max_reward on SFX. it's correct

Reply:
ahh ok, my bad.